packages/flutter/test/material/segmented_button_test.dart

```diff
@@ -1218,6 +1218,103 @@ void main() {
     await tester.pumpWidget(buildButton(enabled: false));
     expect(iconStyle(tester, Icons.add).color, disabledIconColor);
   });
+
+  testWidgets('SegmentedButton border sides respects states', (WidgetTester tester) async {
+    const Color disabledColor = Color(0XFF999999);
+    const Color hoveredColor = Color(0XFF0000FF);
+    const Color focusedColor = Color(0XFF00FF00);
+    const Color selectedColor = Color(0XF0001234);
+    const Color enabledColor = Color(0XFFFF0000);
+
+    Widget buildButton({bool enabled = true, WidgetStateProperty<BorderSide?>? side}) {
+      return MaterialApp(
+        home: Material(
+          child: Center(
+            child: SegmentedButton<int>(
+              style: ButtonStyle(side: side),
+              segments: const <ButtonSegment<int>>[
+                ButtonSegment<int>(value: 0, label: Text('Add'), icon: Icon(Icons.add)),
+                ButtonSegment<int>(value: 1, label: Text('Subtract'), icon: Icon(Icons.remove)),
+                ButtonSegment<int>(
+                  value: 2,
+                  label: Text('Multiply'),
+                  icon: Icon(Icons.multiple_stop),
+                ),
+              ],
+              showSelectedIcon: false,
+              onSelectionChanged: enabled ? (Set<int> selected) {} : null,
+              selected: const <int>{1},
+            ),
+          ),
+        ),
+      );
+    }
+
+    await tester.pumpWidget(
+      buildButton(
+        side: WidgetStateProperty.resolveWith((Set<WidgetState> states) {
+          if (states.contains(WidgetState.hovered)) {
```

Comment:
Can we also test when it's both hovered and selected, what the border will look like? Also for the case when both focused and selected:)

Reply:
Added these checks. Also refactored to use `fromMap`. I always forget it exist. :D